test/feature/steps/init.js

```diff
@@ -0,0 +1,58 @@
+const { Given, When, Then, Before, After } = require('cucumber');
+const Preferences = require('preferences');
+const { exec } = require('child_process');
+const chai = require('chai');
+const sinon = require('sinon');
+const sinonChai = require('sinon-chai');
+const path = require('path');
+const figlet = require('figlet');
+const chalk = require('chalk');
+const controller = require('../../../lib/controller');
+
+chai.use(sinonChai);
+chai.should();
+
+let username, password, logSpy;
+
+Before(() => {
+  logSpy = sinon.spy(console, 'log');
+});
+
+After(() => {
+  logSpy.restore();
+});
+
+Given('a valid username as {string} and corresponding password as {string}', (user, pass) => {
+  username = user;
+  password = pass;
+});
+
+When('I run init command with {string} as username and {string} as password using {string}', (username, password, inputType, done) => {
+  if(inputType === 'flags') {
+      exec(`autolabjs init -u ${username} -p ${password}`, (err, stdout, stderr) => {
+        done();
+      });
+  }
+  else if(inputType === 'prompt') {
+      const stdin = require('mock-stdin').stdin();
+      process.argv = [ '/usr/local/nodejs/bin/node',
+        '/usr/local/nodejs/bin/autolab', 'init' ];
+      controller.start();
+      setTimeout(() => stdin.send(`${username}\n`), 1);
```

Comment:
can this be done using async / await? If not, what needs to change in order to make this async / await?

Reply:
I am simulating the user input here and I think there is no other way to implement this.

Reply:
@AnkshitJain any ideas for making this statement async / await?

Reply:
async/await makes more sense when we are trying to replace callbacks or native promises. Here, we are trying to add a delay to the input of the init command. async/await will not help here.

Reply:
@AnkshitJain The intent is to give input in an orderly manner, not to give inputs precisely at 1sec, 2sec and 3sec. We can create the same effect by having three nested callbacks which can be represented as async / awaits. I though that such a code structure inherently more suitable than giving precise times for giving inputs to mock stdin.

Reply:
I agree that precise timings is not what we want. If there is a promise version of the method stdin.send we can use that instead of setTimeout. The corresponding function becomes async and then we can have something like 
```
   await stdin.send(`${username}\n`);
   await stdin.send(`${password}\n`);
```     
Again this is possible when we have the send method returning a promise/ being an async method.

Reply:
The `mock-stdin` does not return promises, but we can [promisify](http://bluebirdjs.com/docs/api/promise.promisify.html) the same. The subtle problem is in executing the next line; after executing  `stdin.send(`${username}\n`);`, the rest of autolabcli code takes non-zero time to complete its execution. One indication of the completion of execution is the status / prompt message sent to the stdout. We can mock the stdout (console.log) to receive the output to terminal, verify the same and only then go to the next step.

If we go to the next step too soon, we risk giving some of the input to the previous variable. In this particular case, there is a possibility of a part of the password being given to username field.

Given that all the feature tests essentially require a tango between input and output, which is impossible to get right all the time. Hence, it is better to retain the existing code structure.